src/Compilers/Core/Portable/StrongName/CryptoBlobParser.cs

```diff
@@ -155,156 +143,94 @@ private static unsafe bool IsValidPublicKeyUnsafe(ImmutableArray<byte> blob)
-            var exponent = rsa->PubExp;
-            var modulus = new byte[modulusBitLength >> 3];
-
-            if (blobLen - s_offsetToKeyData < modulus.Length)
-            {
-                publicKey = ImmutableArray<byte>.Empty;
-                return false;
-            }
+            var w = new BlobWriter(3 * sizeof(uint) + s_offsetToKeyData + pubKeyData.Length);
+            w.WriteUInt32(AlgorithmId.RsaSign);
+            w.WriteUInt32(AlgorithmId.Sha);
+            w.WriteUInt32((uint)(s_offsetToKeyData + pubKeyData.Length));
 
-            Marshal.Copy((IntPtr)(blob + s_offsetToKeyData), modulus, 0, modulus.Length);
+            w.WriteByte(type);
+            w.WriteByte(version);
+            w.WriteUInt16(reserved);
+            w.WriteUInt32(algId);
 
-            var newHeader = new BlobHeader()
-            {
-                Type = PublicKeyBlobId,
-                Version = version,
-                Reserved = 0,
-                AlgId = AlgorithmId.RsaSign
-            };
+            w.WriteUInt32(magic);
+            w.WriteUInt32(bitLen);
+            w.WriteUInt32(pubExp);
 
-            var newRsaKey = new RsaPubKey()
-            {
-                Magic = RSA1, // Public key
-                BitLen = modulusBitLength,
-                PubExp = exponent
-            };
+            w.WriteBytes(pubKeyData);
 
-            publicKey = CreateSnPublicKeyBlob(newHeader, newRsaKey, modulus);
-            return true;
+            return w.ToImmutableArray();
         }
 
         /// <summary>
         /// Try to retrieve the public key from a crypto blob.
         /// </summary>
         /// <remarks>
-        /// Can be either a PUBLICKEYBLOB or PRIVATEKEYBLOB. The BLOB must /// be unencrypted.
+        /// Can be either a PUBLICKEYBLOB or PRIVATEKEYBLOB. The BLOB must be unencrypted.
         /// </remarks>
         public unsafe static bool TryGetPublicKey(ImmutableArray<byte> blob, out ImmutableArray<byte> publicKey)
         {
-            publicKey = ImmutableArray<byte>.Empty;
-
             // Is this already a strong name PublicKeyBlob?
             if (IsValidPublicKey(blob))
             {
                 publicKey = blob;
                 return true;
             }
 
+            publicKey = ImmutableArray<byte>.Empty;
+
             // Must be at least as large as header + RSA info
-            if (blob.Length < sizeof(BlobHeader) + sizeof(RsaPubKey))
+            if (blob.Length < BlobHeaderSize + RsaPubKeySize)
             {
                 return false;
             }
 
-            fixed (byte* backing = blob.DangerousGetUnderlyingArray())
+            fixed (byte* ptr = blob.DangerousGetUnderlyingArray())
             {
-                var header = (BlobHeader*)backing;
-                var rsa = (RsaPubKey*)(backing + sizeof(BlobHeader));
+                var blobReader = new BlobReader(ptr, blob.Length);
+
+                var type = blobReader.ReadByte();
```

Comment:
Why do you need to know about any structures? The structs are artificial. That said, adding a couple of comments  like
// header
...

//  public key
...

would be good.

---
In reply to: [129946254](https://github.com/dotnet/roslyn/pull/21072#discussion_r129946254) [](ancestors = 129946254)